docs/rfcs/2024-08-06-json-datatype.md

```diff
@@ -0,0 +1,141 @@
+---
+Feature Name: Json Datatype
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/4230
+Date: 2024-8-6
+Author: "Yuhan Wang <profsyb@gmail.com>"
+---
+
+# Summary
+This RFC proposes a method for storing and querying JSON data in the database.
+
+# Motivation
+JSON is widely used across various scenarios. Direct support for writing and querying JSON can significantly enhance the database's flexibility.
+
+# Details
+
+## User Interface
+The feature introduces a new data type for the database, similar to the common JSON type. Data is written as JSON strings and can be queried using functions.
+
+For example:
+```SQL
+CREATE TABLE IF NOT EXISTS test (
+    ts TIMESTAMP TIME INDEX,
+    a INT,
+    b JSON
```

Comment:
I'm not sure if the new `JSON` type is good. It erases the detailed type info and makes everything in this field dynamic. Is it possible to represent this in schema with a clear `Struct` type?

We can still invoke those JSON-related functions on `String` or `Bytes` fields. But not include a new opaque `JSON` type. If we have some optimization to do for `JSON` column, defining this as a column constraint also looks viable. E.g.:
```sql
CREATE TABLE IF NOT EXISTS test (
    ts TIMESTAMP TIME INDEX,
    b BYTES CHECK (is_json)
```

Reply:
> Is it possible to represent this in schema with a clear Struct 

However, the question here is why not let users use PIPELINE to extract all fields from the nested struct? I think the 
dynamic is key for the `JSON` type.